HostProcess containers with Hypervisor Isolation

rawahars opened the pull request:
### Summary
This PR adds support for running HostProcess containers with hypervisor isolation. These HPCs will run as privileged processes within the UVM. Few key aspects of this feature are-
- This feature will be applicable for WS2025 and later.
- A pod needs to be marked privileged by passing in annotation 'microsoft.com/hostprocess-container'. This ensures that the UVM can support HPCs alongside normal containers.
- Any container which has 'microsoft.com/hostprocess-container' annotation will be launched as HostProcess Container within the UVM.
- Other containers will be launched as normal process-isolated containers within the UVM.
- GCS will create the HPCs based on the specs. The behavior is similar to existing HPCs for process-isolated cases.
- Annotations for Inherit user will lead to the HPC running as System user within UVM.
- Annotation for custom rootfs will lead to the container filesystem being virtualized in a directory with that custom name. If not provided, `C:\hpc` is used by default.

### Primary changes

* Refactored pod sandbox logic to use a centralized `updateConfigForHostProcessContainer` method for handling privileged container configuration and annotation passthrough. This method enforces correct sandbox/container combinations and properly sets user and annotation values for HostProcess containers.
* Added new fields to the container and storage schema (`ContainerType` and `ContainerRootPath`) to support HostProcess container identification and custom root filesystem locations. These are set during container creation when an isolated job container is detected.
* Introduced `handleProcessArgsForIsolatedJobContainer` to ensure that isolated job containers have their process arguments correctly prefixed with `cmd /c` when necessary, supporting both `CommandLine` and `Args` fields. This logic is invoked during container and exec creation.
* Improved container creation flow to distinguish between job containers running on the host and those inside a UVM, ensuring the right creation path is used.

### Testing Enhancements

* Added comprehensive unit tests for both HostProcess container configuration updates and process argument handling, covering a wide range of scenarios and edge cases to ensure robust behavior.
* Tests have been added to cover most of the newly added logic.

rawahars committed:
Added support for hypervisor isolated hpc

This commit adds support for running HostProcess containers with hypervisor isolated workloads. Therefore, these HPCs will run as privileged processes within the UVM. Some of the key aspects of this feature are-
- A pod needs to be marked privileged by passing in annotation 'microsoft.com/hostprocess-container'
- Any container which has 'microsoft.com/hostprocess-container' annotation will be launched as HostProcess Container.
- Other containers will be launched as normal process-isolated containers within the UVM
- HPC within UVM will not have access to any network since the sandbox (UVM) does not have any network.
- GCS will create the HPCs based on the specs. The behaviour is similar to existing HPCs for process-isolated case.
- Annotations for Inherit user will lead to the HPC running as System user within UVM.
- Annotation for custom rootfs will lead to the container filesystem being virtualized in a directory with that custom name. If not provided, `C:\hpc` is used by default.

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

rawahars committed:
Added unit tests for HostProcess Container helper methods

This commit adds thorough unit tests for:
- IsJobContainer
- IsIsolatedJobContainer

These tests explicitly verify the expected behavior for
Windows HostProcess containers across isolation modes and OS targets.

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

rawahars committed:
Added unit tests for updateConfigForHostProcessContainer method

This commit adds thorough unit tests for the updateConfigForHostProcessContainer method. We test the following functionality-

- Reject privileged container in unprivileged sandbox (isolated HPC) — return error when container requests microsoft.com/hostprocess-container=true but the pod lacks it.
- Allow privileged container in privileged sandbox (isolated HPC) — ensure no unexpected mutations when no passthrough annotations are present.
- Block normal process-isolated container in process-isolated HPC pod — enforce constraint that all containers in a process HPC pod must also be job containers (error on mixing).
- Allow normal hypervisor-isolated container in hypervisor-isolated HPC pod.
- Passthrough annotations to privileged containers — propagate HostProcessInheritUser and HostProcessRootfsLocation from pod → container for: hypervisor-isolated HPC pods with privileged containers, and process-isolated HPC pods with privileged containers.
-  No passthrough for normal containers — verify annotations aren’t propagated when the container is non-privileged.
- Set SYSTEM user for hypervisor-isolated privileged containers — when HostProcessInheritUser=true, set Process.User.Username to NT AUTHORITY\SYSTEM.
- Do not change user for normal containers — ensure container user remains unchanged even if the pod has HostProcessInheritUser=true.
- Force HostProcessInheritUser to "false" for normal containers — in hypervisor-isolated privileged pods, if a non-privileged container sets the inherit annotation, flip it to "false".

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

rawahars committed:
Added unit tests for handleProcessArgsForIsolatedJobContainer

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

jterry75 commented on internal/hcsoci/hcsdoc_wcow.go:
```diff
@@ -490,6 +490,16 @@ func createWindowsContainerDocument(ctx context.Context, coi *createOptionsInter
 	}
+
+	if oci.IsIsolatedJobContainer(coi.Spec) {
+		v2Container.ContainerType = "HostProcess"
```
This isnt a valid ContainerType in the product code. I dont think we should reuse it like this unless we get a host side change that this becomes a valid value

rawahars replied:
`ContainerType` is being added to the HCS V2 schema as part of the effort to support HostProcess containers.
This PR builds on top of the other changes on HCS/GCS side.

rawahars replied:
Sent you the PR offline.

rawahars committed:
Rectified the schema for custom rootfs for HCS HPCs

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

rawahars committed:
Rectified the logic for user inherit functionality for HPCs

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

jterry75 commented on internal/oci/util.go:
```diff
@@ -21,6 +21,19 @@ func IsIsolated(s *specs.Spec) bool {
+
+// IsIsolatedJobContainer checks if `s` is asking for a Windows job container.
+// This request is for running HPC within guest.
+func IsIsolatedJobContainer(s *specs.Spec) bool {
```
I'd prefer we stick with the above. IsJobContainer returns true when Linux == nil && Windows != nil && annotation[HPC] == true

IsIsolated is already true above based on setting HyperV.

So why can't you just do if IsJobContainer(s) && IsIsolated(s) of you want both?

rawahars replied:
Made the suggested change

jterry75 commented on cmd/containerd-shim-runhcs-v1/task_hcs.go:
```diff
@@ -209,6 +211,12 @@ func newHcsTask(
+	// For Isolated Job containers, we need special handling wherein if the command line
+	// is not specified, then we add 'cmd /C' by default.
+	if oci.IsIsolatedJobContainer(s) {
+		handleProcessArgsForIsolatedJobContainer(s, s.Process)
```
Why is this part of it special for HPC? I dont get it, this is true for all Windows containers right? At this point its just another container

rawahars replied:
It's specific to HostProcess containers because of **when** the process is joined to the silo.

For a normal WCOW container (process- or Hyper-V-isolated), the guest creates the workload process from inside the container silo, so `CreateProcess` resolves the image name and `PATH` using the silo's filesystem view — the bind-mounted rootfs, the container's `System32`, etc. A bare echo / dir or a binary that lives only in the rootfs works fine because the silo is already in scope at creation time.

For a HostProcess container the process is created outside the silo and only assigned to the silo's job object after the fact. That means image-name resolution happens against the launcher's view (host/UVM System32, Windows, host PATH), not the container's rootfs. Shell builtins (echo, set, redirection, …) and any rootfs-only executable would fail with ERROR_FILE_NOT_FOUND.

Wrapping with `cmd /c` sidesteps this: `cmd.exe` is guaranteed to exist under the launcher's `System32`, and once it's joined to the silo it runs inside the container's filesystem view, where it can resolve builtins and rootfs-relative paths normally. Regular WCOW containers don't need this because they never have that `"created outside, joined after"` gap.

rawahars commented on internal/hcs/schema2/storage.go:
```diff
@@ -18,4 +18,7 @@ type Storage struct {
 	QoS *StorageQoS `json:"QoS,omitempty"`
+
+	// Path to the root of the container's filesystem. This is useful in case of HostProcess containers where the container rootfs is different from C:\.
+	PrivilegedContainerRootPath string `json:"PrivilegedContainerRootPath,omitempty"`
```
Sent you the PR for HCS side changes.

rawahars commented on cmd/containerd-shim-runhcs-v1/task_hcs_test.go:
```diff
@@ -318,3 +321,155 @@ func Test_hcsTask_DeleteExec_2ndExecID_ExitedState_Success(t *testing.T) {
+			taskSpec := &specs.Spec{Annotations: tt.taskAnnotations}
+			// Ensure initial Username is set if provided
+			if tt.initialUsername != "" {
+				tt.specs.User.Username = tt.initialUsername
```
Nice catch! I had initially planned on test cases to change the usernames via this parameter. I will remove this in the revision.

rawahars committed:
Addressed PR feedback

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

rawahars committed:
Schema changes inline with HCS

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

rawahars committed:
review comments 2

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

rawahars committed:
reject task creation if the UVM does not support HPCs

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

rawahars committed:
allow HPCs in any UVM

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

rawahars committed:
HPC: tighten inherit-user scoping, harden cmd-wrapping, and fix path sanitizer

- pod: clear HostProcessInheritUser on all non-HPC containers.
- task_hcs: make handleProcessArgsForIsolatedJobContainer nil-safe and
  detect cmd/cmd.exe via filepath.Base so absolute paths are recognised
  and tools like `cmdkey` aren't mis-identified as cmd.
- ospath.Sanitize: compare disallowed prefixes case-insensitively with a
  path-separator boundary (C:\WindowsBackup no longer matches C:\Windows),
  and drop the os.Lstat "already exists" check.
- uvm.HostProcessContainerSupported: guard against a nil guest connection
  and use IsHostProcessContainerSupported().
- Update tests to cover the new behaviour.

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

marma-dev commented on internal/ospath/path.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+// Sanitize validates and normalizes a Windows path.
+func Sanitize(path string, disallowedPrefixes []string) (string, error) {
```
Have we thought about handling/sanitizing relative paths? 

A value like `..\..\Windows\System32` passes through `filepath.Clean` unchanged (since it's already canonical as a relative path) and won't match any disallowed prefix check because those are all absolute 

In case the GCS resolves `PrivilegedContainerRootPath` relative to a base directory, would this allow escaping into protected locations?

rawahars replied:
Thanks for pointing that out. I have fixed it in the latest revision.

marma-dev commented on internal/hcs/schema2/container.go:
```diff
@@ -9,6 +9,19 @@
+const (
+	// IsolationTypeProcess is a fully process-isolated, namespace-isolated
+	// Windows Server container.
+	IsolationTypeProcess IsolationType = "Process"
```
Nit: This is currently unreferenced

marma-dev commented on cmd/containerd-shim-runhcs-v1/task_hcs.go:
```diff
@@ -1015,6 +1027,55 @@ func (ht *hcsTask) requestAddContainerMount(ctx context.Context, resourcePath st
+		if base != "cmd" && base != "cmd.exe" {
+			p.Args = append([]string{"cmd", "/c"}, p.Args...)
+		}
+	}
```
Could there be a case where a process spec has both `CommandLine` and `Args` populated? 

If so, then both get [cmd /c] prepended independently. 

IIRC, CommandLine takes precedence — so modifying `Args` in that case is technically unnecessary. It's not harmful but could confuse debugging.

rawahars replied:
Thanks for catching this! Fixed it in the current revision.

marma-dev reviewed (approved):
Added some minor comments. 

Overall, the PR looks good!

rawahars committed:
address review comments 2

Signed-off-by: Harsh Rawat <harshrawat@microsoft.com>

marma-dev reviewed (approved):
LGTM
